flink/v1.12/flink/src/test/java/com/netease/arctic/flink/TestArcticCatalogTablePartitions.java

```diff
@@ -0,0 +1,131 @@
+
+import com.netease.arctic.BasicTableTestHelper;
+import com.netease.arctic.ams.api.TableFormat;
+import com.netease.arctic.catalog.BasicCatalogTestHelper;
+import com.netease.arctic.flink.catalog.ArcticCatalog;
+import com.netease.arctic.flink.util.DataUtil;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableMap;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.junit.Assert;
+import org.junit.Test;
+
+import java.util.LinkedList;
+import java.util.List;
+
+import static org.apache.flink.table.api.Expressions.$;
+
+public class TestArcticCatalogTablePartitions extends FlinkTestBase {
+  private final String tableName = "test_partition_table";
+  private final String db = "test_partition_db";
+
+  public TestArcticCatalogTablePartitions() {
+    super(new BasicCatalogTestHelper(TableFormat.MIXED_ICEBERG), new BasicTableTestHelper(true, true));
+  }
+
+  public void before() throws Exception {
+    super.before();
+    super.config();
+  }
+
+  @Test
+  public void testListPartitionsUnKeyedTable() throws TableNotPartitionedException {
+    List<Object[]> data = new LinkedList<>();
+    data.add(new Object[]{1, "mark", "2023-10-01"});
+    data.add(new Object[]{2, "Gerry", "2023-10-02"});
+
+    List<ApiExpression> rows = DataUtil.toRows(data);
+    Table input = getTableEnv().fromValues(DataTypes.ROW(
+            DataTypes.FIELD("id", DataTypes.INT()),
+            DataTypes.FIELD("name", DataTypes.STRING()),
+            DataTypes.FIELD("dt", DataTypes.STRING())
+        ),
+        rows
+    );
+    getTableEnv().createTemporaryView("input", input);
+
+    sql("CREATE CATALOG arcticCatalog WITH %s", toWithClause(props));
+
+    sql("CREATE TABLE IF NOT EXISTS arcticCatalog." + db + "." + tableName + "(" +
+        " id INT, name STRING, dt STRING) PARTITIONED BY (dt)");
+
+    sql("INSERT INTO %s select * from input", "arcticCatalog." + db + "." + tableName);
+    ObjectPath objectPath = new ObjectPath(db, tableName);
+    ArcticCatalog arcticCatalog = (ArcticCatalog) getTableEnv().getCatalog("arcticCatalog").get();
+    List<CatalogPartitionSpec> list = arcticCatalog.listPartitions(objectPath);
+
+
+    List<CatalogPartitionSpec> expected = Lists.newArrayList();
+    CatalogPartitionSpec partitionSpec1 = new CatalogPartitionSpec(ImmutableMap.of("dt", "2023-10-01"));
+    CatalogPartitionSpec partitionSpec2 = new CatalogPartitionSpec(ImmutableMap.of("dt", "2023-10-02"));
+    expected.add(partitionSpec1);
+    expected.add(partitionSpec2);
+    Assert.assertEquals("Should produce the expected catalog partition specs.", list, expected);
+  }
+
+
+  @Test
+  public void testListPartitionsKeyedTable() throws TableNotPartitionedException {
+    List<Object[]> data = new LinkedList<>();
+    data.add(new Object[]{1, "mark", "2023-10-01"});
+    data.add(new Object[]{2, "Gerry", "2023-10-02"});
+    data.add(new Object[]{RowKind.DELETE, 2, "Gerry", "2023-10-02"});
```

Comment:
Like this, first insert a piece of data with ID 2 and then delete it. ListPartition cannot determine whether the partition  `2023-10-02` really exists.
@YesOrNo828 
 WDYT?

Reply:
@huyuanfeng2018 I don't think there's an excellent way to return the data's partitions. Additionally, the issue is present in the iceberg v2 table.
Perhaps we can add a comment to the method to clarify that the partitions returned in this scenario encompass all the partitions related to writing the data, including -D and -U.

cc @zhoujinsong  WDYT?

Reply:
I think it's okay to return this partition in this situation.